perceval/backends/mediawiki.py

```diff
@@ -0,0 +1,475 @@
+
+from ..backend import Backend, BackendCommand, metadata
+from ..cache import Cache
+from ..errors import BackendError, CacheError
+from ..utils import (DEFAULT_DATETIME,
+                     datetime_to_utc,
+                     str_to_datetime,
+                     urljoin)
+
+
+logger = logging.getLogger(__name__)
+
+MAX_RECENT_DAYS = 30  # max number of days included in MediaWiki recent changes
+
+
+class MediaWiki(Backend):
+    """MediaWiki backend for Perceval.
+
+    This class retrieves the wiki pages and edits from a
+    MediaWiki site. To initialize this class the
+    URL must be provided.
+
+    :param url: MediaWiki url
+    :param cache: cache object to store raw data
+    :param origin: identifier of the repository; when `None` or an
+        empty string are given, it will be set to `url` value
+    """
+    version = '0.1.0'
+
+    def __init__(self, url, cache=None, origin=None):
+        origin = origin if origin else url
+
+        super().__init__(origin, cache=cache)
+        self.url = url
+        self.client = MediaWikiClient(url)
+
+    def __get_max_date(self, reviews):
+        """" Get the max date in unixtime format from reviews """
+        max_ts = 0
+        for review in reviews:
+            ts = str_to_datetime(review['timestamp'])
+            ts = ts.replace(tzinfo=dateutil.tz.tzutc())
+            if ts.timestamp() > max_ts:
+                max_ts = ts.timestamp()
+        return max_ts
+
+    @metadata
+    def fetch(self, from_date=DEFAULT_DATETIME):
+        """Fetch the pages from the backend url.
+
+        The method retrieves, from a MediaWiki url, the
+        wiki pages.
+
+        :returns: a generator of pages
+        """
+
+        self._purge_cache_queue()
+
+        if from_date == DEFAULT_DATETIME:
+            from_date = None
+        else:
+            from_date = datetime_to_utc(from_date).replace(tzinfo=None)
+
+        mediawiki_version = self.client.get_version()
+        logging.info("MediaWiki version: %s", mediawiki_version)
+
+        if mediawiki_version[0] >= 1 and mediawiki_version[1] >= 27:
+            fetcher = self.__fetch_1_27(from_date)
+        else:
+            fetcher = self.__fetch_pre1_27(from_date)
+
+        for page_reviews in fetcher:
+            yield page_reviews
+
+
+    def __fetch_1_27(self, from_date=None):
+        """Fetch the pages from the backend url for MediaWiki >=1.27
+
+        The method retrieves, from a MediaWiki url, the
+        wiki pages.
+
+        :returns: a generator of pages
+        """
+
+        logger.info("Looking for pages at url '%s'", self.url)
+
+        self._purge_cache_queue()
+        npages = 0  # number of pages processed
+
+        # The pages are organized in content namespaces
+        raw_namespaces = self.client.get_namespaces()
+        namespaces = json.loads(raw_namespaces)["query"]["namespaces"]
+        namespaces_contents = [ns for ns in namespaces if 'content' in namespaces[ns].keys()]
```

Comment:
I think you should explain why you do manage the namespaces this way and why some are filtered and some others not.